### PR TITLE
Activity Events

### LINKED ISSUE #754: Figure out optimal Dockerfile
Current setup uses a `volume` to transfer code over into the container

So when we run npm install on build, the modules get installed. But the volume then overrides the `node_modules` folder with nothing.

The temp solution for now is to run `npm install` locally.

Ideally we should not require this step.

### PR DESCRIPTION
**Pull request by deadlybutter**
#### What's this PR do?
- Produces `signup` and `reportback` events to Rabbit MQ  `activity events` queue
- Updates docker compose setup to add RabbitMQ container
 - Also includes changes from https://github.com/DoSomething/gambit/pull/755 so I wasn't blocked from working.

#### How should this be reviewed?
Post to Gambit locally & check the events are fired

#### Any background context you want to provide?
Needed for relative reminder messaging

#### Relevant tickets
Fixes #754 
Touches #742

### CODE REVIEW

File: lib/rabbit.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+}
+
+function makeConnection() {
+  logger.info('Attempting amqp connection');
+
+  const amqp = require('amqplib');
+
+  amqp.connect(process.env.RABBIT_URI)
+  .then(connection => connection.createChannel())
+  .then(ch => {
```
Comment (sergii-tkachenko) [inline lib/rabbit.js]: Should be a oneliner.

File: lib/rabbit.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    channel = ch;
+  })
+  .catch(err => {
```
Comment (sergii-tkachenko) [inline lib/rabbit.js]: Parentheses.

File: lib/rabbit.js [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    onError(err);
+    setTimeout(makeConnection, 2000);
```
Comment (sergii-tkachenko) [inline lib/rabbit.js]: Why 2000?

File: lib/rabbit.js [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  });
+}
+makeConnection();
+
+module.exports = {
+  produce(queueName, message) {
+    const outgoing = JSON.stringify(message);
+
+    return channel
```
Comment (sergii-tkachenko) [inline lib/rabbit.js]: Not readable.

File: lib/rabbit.js [hunk 5]
```diff
@@ -0,0 +1,39 @@
+'use strict';
+
+const stathat = app.locals.stathat;
+const logger = app.locals.logger;
+
+let channel = {};
+
+function onError(err) {
+  logger.error('Rabbit Error', err);
+  stathat(`gambit-core rabbit error ${err.code}`);
```
Comment (sergii-tkachenko) [inline lib/rabbit.js]: What if it's not an exception? You need to handle err.code = undefined.

File: lib/rabbit.js (hunk 1 above)
Comment (sergii-tkachenko) [inline lib/rabbit.js]: Also, `ch` is not informative.

File: lib/rabbit.js [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+  .then(connection => connection.createChannel())
+  .then(ch => {
+    channel = ch;
+  })
+  .catch(err => {
+    onError(err);
+    setTimeout(makeConnection, 2000);
+  });
+}
+makeConnection();
```
Comment (sergii-tkachenko) [inline lib/rabbit.js]: This is bad. Really difficult to notice. Make this a class with a constructor.

File: lib/rabbit.js (hunk 1 above)
Comment (deadlybutter) [inline lib/rabbit.js]: Yeah it was a bit hard to name this because I couldn't use `channel`. Suggestions?

File: lib/rabbit.js (hunk 3 above)
Comment (deadlybutter) [inline lib/rabbit.js]: No specific reason, just felt like 2 seconds was a reasonable retry timeout

File: lib/rabbit.js (hunk 4 above)
Comment (deadlybutter) [inline lib/rabbit.js]: What is not readable?

File: lib/rabbit.js (hunk 1 above)
Comment (sergii-tkachenko) [inline lib/rabbit.js]: Refactor this to a class. So channel will be `this.channel`.

File: lib/rabbit.js (hunk 3 above)
Comment (sergii-tkachenko) [inline lib/rabbit.js]: This should be in a constant then.

**Comment by deadlybutter** [comment]
updated but not tested, also seems to have lint problem

File: lib/rabbit.js [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+      this.channel = channel;
+    })
+    .catch((err) => {
+      onError(err);
+      setTimeout(this.connect, AMQP_RETRY_FREQUENCY);
+    });
+  }
+}
+
+const rabbitConnection = new Connection();
```
Comment (sergii-tkachenko) [inline lib/rabbit.js]: I meant you can just put everything in the class and export it.

File: app/controllers/CampaignBotController.js [hunk 8]
```diff
@@ -179,6 +180,13 @@ class CampaignBotController {
   postReportback(req) {
     this.debug(req, 'postReportback');
 
+    rabbit.produce('activity events', {
+      type: 'reportback',
+      userId: req.user._id,
+      campaignId: req.campaign._id,
+      campaignRunId: req.campaign.current_run,
+    });
```
Comment (deadlybutter) [inline app/controllers/CampaignBotController.js]: @aaronschachter I did not see that issue, thanks for pointing that out! I'll move this over to Signup then

**Comment by deadlybutter** [comment]
@sergii-tkachenko sweet, i think this is much better now

**Comment by aaronschachter** [comment]
@deadlybutter I'm not following the logic in this branch, and there's no corresponding issue to explain what's going on here. Why are we posting `signup` and `reportback` events?

**Comment by deadlybutter** [comment]
@aaronschachter By posting events to a message broker we can more easily spin up new micro services (in contrast to adding more uri/keys everywhere & relying on HTTP). In this case, the next step is a micro service to keep track of sending relative reminder messages. It will consume these activity events to figure out when it should be posting reminders. Sorry I haven't documented any of this anywhere yet in Github or Trello, all over the place atm but I will make sure to get that done for you!

**Comment by aaronschachter** [comment]
We're posting Signups and Reportbacks to the Phoenix API -- which posts to Quicksilver, so isn't posting events to a message broker already taken care of?  Is the intent to implement something like https://github.com/DoSomething/quicksilver-api/issues/98?

I know I'm on leave, so take this with a grain of salt -- but my $0.02 on this one is to hold off on any more development (and hold on merging) until there's a full technical spec that documents what the intent of this work is for, and how it will be implemented.

**Comment by deadlybutter** [comment]
@sergii-tkachenko Im going to reopen the other PR that had the Docker fixes from this PR, as those should still be merged in anyway

@aaronschachter ok, update! I talked w/ Sergii for a bit -- we're both still onboard with _moving_ Gambit towards a more distributed system versus a single application. I think we've talked about this before but I would be happy to write out the reasoning if im mistaken!

However we both agreed to minimize the scope as much as possible. For now, since Quicksilver already has signup/reportback events, we're just going to tap into that system. We're going to add a signup/reportback reminder message to the campaign object & expose that in the API. From there Quicksilver can use that to send out reminders. Let me know if you have any more questions!

**Comment by aaronschachter** [comment]
@deadlybutter I do think we need at least a Github issue to serve as the spec for why distributed system and specifically when/where we'll use it.

Using Quicksilver for sending relative reminders definitely makes sense -- @sergii-tkachenko and I had mentioned it either should belong in QS or the source system of Signups (which would be Phoenix... but eventually Rogue.. so definitely makes sense for QS to handle to unblock development)

**Pull request closed without merging**